### PR TITLE
Issue #8765: implement missing methods on PG Range

### PR DESCRIPTION
**Pull request by lelit**
### Description
This PR implements missing methods on the PG `Range` class, as described by issue #8765.

### Checklist
This pull request is:

- [ ] A documentation / typographical error fix
- [ ] A short code fix
- [x] A new feature implementation

### CODE REVIEW

**Comment by lelit** [comment]
Note that I re-used the existing test fixtures, originally designed to exercise the _containment_ methods, also in the new comparison methods tests. This is not ideal, because several corner cases are missing, in particular for the `adjacent_to()` method. I will either expand those fixtures or distill new ones with more appropriate ranges.

**Comment by lelit** [comment]
For completeness, this is the (a bit redundant) test I used to ensure `adjacent_to()` is working on those cases:
```diff
diff --git a/test/dialect/postgresql/test_dialect.py b/test/dialect/postgresql/test_dialect.py
index 27d4a4cf9..aac1a838c 100644
--- a/test/dialect/postgresql/test_dialect.py
+++ b/test/dialect/postgresql/test_dialect.py
@@ -4,6 +4,7 @@
 import logging
 import logging.handlers
 
+from sqlalchemy import literal_column
 from sqlalchemy import BigInteger
 from sqlalchemy import bindparam
 from sqlalchemy import cast
@@ -30,7 +31,9 @@
 from sqlalchemy import TypeDecorator
 from sqlalchemy.dialects.postgresql import base as postgresql
 from sqlalchemy.dialects.postgresql import HSTORE
+from sqlalchemy.dialects.postgresql import INT4RANGE
 from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.dialects.postgresql import NUMRANGE
 from sqlalchemy.dialects.postgresql import psycopg as psycopg_dialect
 from sqlalchemy.dialects.postgresql import psycopg2 as psycopg2_dialect
 from sqlalchemy.dialects.postgresql import Range
@@ -1210,3 +1213,92 @@ def test_get_dialect(self):
     def test_async_version(self):
         e = create_engine("postgresql+psycopg_async://")
         is_true(isinstance(e.dialect, psycopg_dialect.PGDialectAsync_psycopg))
+
+
+class TestRange(fixtures.TestBase):
+    __only_on__ = "postgresql"
+    __backend__ = True
+
+    @testing.combinations(
+        "empty",
+        "(1,2]",
+        "[1,2)",
+        "(,)",
+        "[,1)",
+        "[1,)",
+        "[1,1]",
+        "[1,4)",
+        "[1,4]",
+        "(1,4]",
+        "(1,4)",
+        argnames="r2repr",
+    )
+    @testing.combinations(
+        "empty",
+        "[,)",
+        "[,1)",
+        "[1,)",
+        "[1,4)",
+        "[1,4]",
+        "(1,4]",
+        "(1,4)",
+        "(2,3]",
+        "[2,3]",
+        "(3,4]",
+        "[3,5)",
+        "[0,0]",
+        "[-1,0)",
+        "[-4,1)",
+        "(6,8)",
+        argnames="r1repr",
+    )
+    def test_int4range_adjacent(self, connection, r1repr, r2repr):
+        q = select(
+            literal_column(f"'{r1repr}'::int4range", INT4RANGE).label("r1"),
+            literal_column(f"'{r2repr}'::int4range", INT4RANGE).label("r2"),
+            literal_column(f"'{r1repr}'::int4range -|- '{r2repr}'::int4range"),
+        )
+        r1, r2, adjacent = connection.execute(q).first()
+        eq_(r1.adjacent_to(r2), adjacent)
+
+    @testing.combinations(
+        "empty",
+        "(1.1,2.9]",
+        "[1.1,2.9)",
+        "(,)",
+        "[,1.1)",
+        "[1.1,)",
+        "[1.1,1.1]",
+        "[1.1,4.5)",
+        "[1.1,4.5]",
+        "(1.1,4.5]",
+        "(1.1,4.5)",
+        argnames="r2repr",
+    )
+    @testing.combinations(
+        "empty",
+        "[,)",
+        "[,1.1)",
+        "[1.1,)",
+        "[1.1,4.5)",
+        "[1.1,4.5]",
+        "(1.1,4.5]",
+        "(1.1,4.5)",
+        "(2.9,3.2]",
+        "[2.9,3.2]",
+        "(3.2,4.5]",
+        "[3.2,5.5)",
+        "[0.1,0.1]",
+        "[-1.1,0.1)",
+        "[-4.5,1.1)",
+        "(6.1,8.1)",
+        argnames="r1repr",
+    )
+    def test_numrange_adjacent(self, connection, r1repr, r2repr):
+        q = select(
+            literal_column(f"'{r1repr}'::numrange", NUMRANGE).label("r1"),
+            literal_column(f"'{r2repr}'::numrange", NUMRANGE).label("r2"),
+            literal_column(f"'{r1repr}'::numrange -|- '{r2repr}'::numrange"),
+        )
+        r1, r2, adjacent = connection.execute(q).first()
+        eq_(r1.adjacent_to(r2), adjacent)
```

**Comment by lelit** [comment]
I'd like to know your opinion on the following:

1. Is there a particular reason behind the choice of [`__add__`()](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py#L263) method, in contrast to what has been done for example with [strictly_right_of](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py#L237) of which `__rshift__` is an alias? For consistency, I'd instead add a (say) `union(self, other)` method, aliased to `__add__`, both on the `Range` class and on the `AbstractRange` one.
2. Was the method [AbstractRange.`__ne__`()](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py#L330) omitted on purpose on `Range`? If not, I'd implement it on the latter as well, again for consistency, with the same consideration above about the naming.
3. If there is a `__ne__`, shouldn't be there also `__eq__`?
4. Given that there is `__add__`, shouldn't be there also `__sub__`?

**Comment by zzzeek** [comment]
> I'd like to know your opinion on the following:
> 
>     1. Is there a particular reason behind the choice of [`__add__`()](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py#L263) method, in contrast to what has been done for example with [strictly_right_of](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py#L237) of which `__rshift__` is an alias? For consistency, I'd instead add a (say) `union(self, other)` method, aliased to `__add__`, both on the `Range` class and on the `AbstractRange` one.

oh you mean why is there no union() method?   by all means we should add one.

> 
>     2. Was the method [AbstractRange.`__ne__`()](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/ranges.py#L330) omitted on purpose on `Range`? If not, I'd implement it on the latter as well, again for consistency, with the same consideration above about the naming.
> 
>     3. If there is a `__ne__`, shouldn't be there also `__eq__`?

`Range` is a dataclass, so ``__eq__()`` and ``__ne__()`` should be implemented already. that was one of the reasons I used dataclasses.   there should be tests for that, if not, we should add them.


> 
>     4. Given that there is `__add__`, shouldn't be there also `__sub__`?

if we have one that works on AbstractRange then yes, if there's a PG operation that this maps towards.

**Comment by lelit** [comment]
Thanks Mike!

On `__add__`, I will then rename existing methods to `union()` method on `Range` and `AbstractRange`, aliasing them to `__add__`.

On `__eq__`/`__ne__`: my question was imprecise, as I knew the dataclass heritage, what I meant was more directed at the different behaviour between `AbstractRange` and `Range`, and whether we care to handle that or not: `Range(1,2,"[]")` is different from `Range(0,3,"()")`, while the respective `Int4Range`s are equal (I mean, `select '[1,2]'::int4range = '(0,3)'::int4range` yields `True`). If we care, I can easily implement the same logic.

And yes, there is a _difference_ operator for ranges in PG (the last one in [this table](https://www.postgresql.org/docs/current/functions-range.html#RANGE-OPERATORS-TABLE)): if get it right, I'll implement that as well, ok?

**Comment by zzzeek** [comment]
> 
> On `__eq__`/`__ne__`: my question was imprecise, as I knew the dataclass heritage, what I meant was more directed at the different behaviour between `AbstractRange` and `Range`, and whether we care to handle that or not: `Range(1,2,"[]")` is different from `Range(0,3,"()")`, while the respective `Int4Range`s are equal (I mean, `select '[1,2]'::int4range = '(0,3)'::int4range` yields `True`). If we care, I can easily implement the same logic.

We probably should!   On my end I'm just trying to build a pipeline, I've not been a user of this particular PostgreSQL feature so I rely upon the users to tell me what non-obvious behaviors it has.  I think making it be as much a mirror of the database type as possible is always best, do you agree?

> 
> And yes, there is a _difference_ operator for ranges in PG (the last one in [this table](https://www.postgresql.org/docs/current/functions-range.html#RANGE-OPERATORS-TABLE)): if get it right, I'll implement that as well, ok?

yup thanks

**Comment by lelit** [comment]
> I think making it be as much a mirror of the database type as possible is always best, do you agree?

Yes, definitely!

**Comment by CaselIT** [comment]
> 1. `union(self, other)` method, aliased to `__add__`, both on the `Range` class and on the `AbstractRange` one.

wouldn't `|` as alias to union be more appropriate? It's what set and dict do. Of we want to follow PG syntax here?

**Comment by CaselIT** [comment]
> PG uses the "+" operator for union(), so we should use `__add__()`

makes sense, IIRC we already have precedence where we opt for the sql syntax vs the python one

**Comment by lelit** [comment]
Also, I think `|` would be slightly misleading: the `+` operator rejects "non-overlapping-non-adjacent" ranges, that is `[1,2] + [3,4]` is ok, while `[1,2] + (3,4]` is not. There does not seem to be an `|` operator that would map to the `range_merge()` function, that would work in that case.

**Comment by lelit** [comment]
I'm going to push the status quo: I implemented the `union()` and the `difference()` methods, but I'm fighting with the test fixtures... I will need to come up with a better scheme :man_shrugging:

**Comment by lelit** [comment]
Uhm, strange: I expected a series of failures... maybe the CI workflow does not exercise the PG specific tests?
Do you have an hint on where to start investigating on the mypy failure? At first, it looks completely unrelated to my changes...

**Comment by CaselIT** [comment]
> Uhm, strange: I expected a series of failures... maybe the CI workflow does not exercise the PG specific tests?

Yes, only sqlite is run on GitHub. Let me bring it on gerrit

> Do you have an hint on where to start investigating on the mypy failure? At first, it looks completely unrelated to my changes...

That's very likely. Mike mentioned that mypy released a new version yesterday that broke things. Don't worry about them for the moment.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision bde39e92290c256ec1baeef9d5975be4e3c7132b of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change bde39e92290c256ec1baeef9d5975be4e3c7132b: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

**Comment by lelit** [comment]
As said, I need a more versatile way of generating different ranges sets to feed the various test functions.

I've come up with the following:

```python
    def _modify_range(
            self,
            r: Range,
            lb: str = None,
            lv: str = None,
            uv: str = None,
            ub: str = None,
            empty=False
    ) -> Range:
        """Take a base range `r` and return an arbitrary transformation.

        The two arguments `lb` and `ub` may be either ``=`` to indicate
        "no change" or one of the bound characters, ``(``, ``[``, ``]``
        and ``)``.

        The two arguments `lv` and `uv` may be either ``∞`` to indicate
        an unbound edge, or an arbitrary expression like ``X + 3*Ɛ``,
        where `X` is the original edge value and `Ɛ` is a small value
        specific to the range type. The expression will be evaluated
        using ``eval()``, with a context containing:

          X
            the original edge value
          L
            the lower edge value
          U
            the upper edge value
          Ɛ
            a small amount of the right type

        When `empty` is True, this simply returns an empty range.
        """

        if empty:
            return Range(None, None, empty=True)
        Ɛ = self._epsilon
        if lb == "=":
            lb = r.bounds[0]
        if ub == "=":
            ub = r.bounds[1]
        if lv == "∞":
            lv = None
        else:
            lv = eval(lv, dict(X=r.lower, L=r.lower, U=r.upper, Ɛ=Ɛ))
        if uv == "∞":
            uv = None
        else:
            uv = eval(uv, dict(X=r.upper, L=r.lower, U=r.upper, Ɛ=Ɛ))
        return Range(lv, uv, bounds=lb + ub)

    @testing.combinations(
        dict(empty=True),
        dict(lb="=", lv="X + Ɛ", uv="X + Ɛ*3", ub="="),
        dict(lb="[", lv="L", uv="L", ub="]"),
        argnames="transform",
    )
    def test_foo(self, transform):
        base = self._data_obj()
        other = self._modify_range(base, **transform)
```

that, given the following mixin:

```python
class _Int4RangeTests:

    _col_type = INT4RANGE
    _col_str = "INT4RANGE"

    def _data_str(self):
        return "[1,4)"

    def _data_obj(self):
        return Range(1, 4)

    _epsilon = 1
```

would generate the following `other` ranges:
```
Range(lower=None, upper=None, bounds='[)', empty=True)
Range(lower=2, upper=7, bounds='[)', empty=False)
Range(lower=1, upper=1, bounds='[]', empty=False)
```

while with the mixin:

```
class _NumRangeTests:

    _col_type = NUMRANGE
    _col_str = "NUMRANGE"

    def _data_str(self):
        return "[1.0,9.0)"

    def _data_obj(self):
        return Range(decimal.Decimal("1.0"), decimal.Decimal("9.0"))

    _epsilon = decimal.Decimal(1)
```

would generate

```
Range(lower=None, upper=None, bounds='[)', empty=True)
Range(lower=Decimal('2.0'), upper=Decimal('12.0'), bounds='[)', empty=False)
Range(lower=Decimal('1.0'), upper=Decimal('1.0'), bounds='[]', empty=False)
```

What's your opinion?

**Comment by CaselIT** [comment]
looking at it seems a bit complex.
Just so I understand correctly, you want a list of parameters that's then used using different kind of ranges with different datatypes, so using something like lambdas where you pass the input range and you return a transformed one can be problematic since the types are different?

**Comment by lelit** [comment]
Basically, yes: Mike already implemented [such mechanism](https://github.com/sqlalchemy/sqlalchemy/blob/main/test/dialect/postgresql/test_types.py#L3879), but it's a bit cumbersome to define those sets with different characteristics for the various test functions.
As an example, this is what I wrote to test the `__eq__()` method, where you want to compare, say, `[1,2]` with `(0,2]` and `(0,3)` and...
```python
    @testing.combinations(
        dict(empty=True),
        dict(lb="[", lv="L", uv="L", ub="]"),
        argnames="r1_transform",
    )
    @testing.combinations(
        dict(empty=True),
        dict(lb="[", lv="L", uv="L", ub="]"),
        dict(lb="[", lv="L", uv="L + Ɛ", ub=")"),
        dict(lb="(", lv="L - Ɛ", uv="L", ub="]"),
        dict(lb="(", lv="L - Ɛ", uv="L + Ɛ", ub=")"),
        argnames="r2_transform",
    )
    def test_equal(self, connection, r1_transform, r2_transform):
        base = self._data_obj()
        r1 = self._modify_range(base, **r1_transform)
        r2 = self._modify_range(base, **r2_transform)

        RANGE = self._col_type
        range_typ = self._col_str

        q = select(
            literal_column(f"'{r1}'::{range_typ} = '{r2}'::{range_typ}")
        )
        equal = connection.execute(q).scalar()
        eq_(
            r1 == r2,
            equal,
            f"{r1} == {r2} != {equal}",
        )
```

Anyway, the last word is obviously yours: if you do not like the above, I will try harder to create different variants of Mike's fixtures.

**Comment by CaselIT** [comment]
using lambda can't this become 
```py
@testing.combinations(
    lambda r, e: Range(None, None, empty=True),
    lambda r, e: Range(r.lower, r.lower, bounds="[]"),
    argnames="r1_transform",
)
@testing.combinations(
    lambda r, e: Range(None, None, empty=True),
    lambda r, e: Range(r.lower, r.lower, bounds="[]"),
    lambda r, e: Range(r.lower, r.lower + e, bounds="[)"),
    lambda r, e: Range(r.lower - e, r.lower, bounds="(]"),
    lambda r, e: Range(r.lower - e, r.lower + e, bounds="()"),
    argnames="r2_transform",
)
```

I find it easier to read.

Are some cases where this would not work?

**Comment by lelit** [comment]
Ok, will do that way, thank you.

**Comment by CaselIT** [comment]
note that it's not set in stone, if some cases require something more complex we can add it. I just like starting simple :)

thanks for working on this!

**Comment by CaselIT** [comment]
Not sure if adding to Range the properties `lb` and `ub` that return the two bounds would be useful?
(or called `lbound`/`ubound` or extended as `lower_bound`/`upper_bound` but at that pind doing `bounds[0]`/`bounds[1]` is probably more convenient)

**Comment by lelit** [comment]
I think this is now ready to be reviewed, the test coverage should be complete now.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 1f930058f9b94d714470ee350baa41d7a6f44b08 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 1f930058f9b94d714470ee350baa41d7a6f44b08 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

**Comment by lelit** [comment]
I saw the failures with asyncpg, but I won't be able to dig into the issue until later this evening: a quick look make me wonder about why the `DataError` exception isn't properly catched and handled for example [here](https://github.com/lelit/sqlalchemy/blob/pg-range%2B%2B/test/dialect/postgresql/test_types.py#L4361), as it happens with psycopgX...

**Comment by lelit** [comment]
I think I got it: performing an _invalid_ range operation under `psycopg` raises a `DataError`, while with `asyncpg` a more generic `DBAPIError`. Not sure if that's a problem somewhere else though.

Wrt the exception, in [those cases I raised](https://github.com/lelit/sqlalchemy/blob/pg-range%2B%2B/lib/sqlalchemy/dialects/postgresql/ranges.py#L466) a generic `ValueError`: does that sound correct?

**Comment by zzzeek** [comment]
hi -

have been swamped w/ major stability types of issues and im also not around tomorrow, so hopefully next week I can look at what's happening here more assuming i can get things calmed down.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 84c1ba21d84a08421a9d18399570d8e962168bd4 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 84c1ba21d84a08421a9d18399570d8e962168bd4 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

File: lib/sqlalchemy/dialects/postgresql/ranges.py [hunk 1]
```diff
@@ -220,52 +284,287 @@ def contains(self, value: Union[_T, Range]) -> bool:
...
+                and supper_b != "]"
+                and self._compare_edges(oupper, rlower_b, supper, supper_b)
+                == 0
+            ):
+                return Range(None, None, empty=True)
+            else:
+                return Range(oupper, supper, bounds=rlower_b + supper_b)
+
+        # TODO: figure out if I handled all the cases above
+        assert False
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/postgresql/ranges.py:565]: **mike bayer** (zzzeek) wrote:

maybe put a message here.  seems like we dont have a good exception class to raise if we got here.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

this looks great!


remaining changes:

1. changelog file

2. .... I can do it in a subsequent commit, I think it's time to move the range tests into test/dialect/postgresql/test_ranges.py


View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

File: lib/sqlalchemy/dialects/postgresql/ranges.py (hunk 1 above)
Comment (lelit) [reply lib/sqlalchemy/dialects/postgresql/ranges.py:565]: Thank you!

> changelog file

Should I try to draft one? If so, it is going to `doc/build/changelog/unreleased_20/GH_ISSUE_NR.rst`, right?

>  seems like we dont have a good exception class to raise if we got here

Would `NotImplementedError("Unhandled this and that case")` fit?

Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/ranges.py:565]: if i understand the case correctly, it means this part of the code should be unreachable, so there would be nothing to "implement", right?   lets just make it "assert False, 'unhandled code path'"

Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/ranges.py:565]: re: changelog file yes, just add one of those in and I can fix it up if it needs..thanks!

Comment (lelit) [reply lib/sqlalchemy/dialects/postgresql/ranges.py]: I did not find a way to check the changelog markup, in particular I'm not sure about the `:attr:` reference that I copied from existing doc.

Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/ranges.py]: no worries, restructured text + sphinx directives took me years to learn, and it's not making things easier we have some of our own conventions buried inside of that

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 21c0df86cc0d1502855527e29425fbffc3f45d64 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 21c0df86cc0d1502855527e29425fbffc3f45d64 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

File: doc/build/changelog/unreleased_20/8765.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+.. change::
+    :tags: usecase, postgresql
+    :tickets: 8765
+
+    Complementing :ticket:`8690`, new comparison methods such as
```
Comment (sqla-tester) [inline doc/build/changelog/unreleased_20/8765.rst:5]: **mike bayer** (zzzeek) wrote:

perfect!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

this is done.  A+ !!!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4182
